pycytominer/cyto_utils/sqlite.py

```diff
@@ -0,0 +1,561 @@
+                """
+            else:
+                # otherwise we will focus on only the column name provided
+                sql_stmt = """
+                SELECT :table_name, name, type, [notnull]
+                FROM pragma_table_info(:table_name) WHERE name = :col_name;
+                """
+
+            # append to column list the results
+            column_list += connection.execute(
+                sql_stmt,
+                {"table_name": str(table[0]), "col_name": str(column_name)},
+            ).fetchall()
+
+    return column_list
+
+
+def contains_conflicting_aff_strg_class(
+    sql_engine: Union[str, Engine],
+    table_name: Optional[str] = None,
+    column_name: Optional[str] = None,
+) -> bool:
+    """
+    Detect conflicting column affinity vs data storage class for
+    entire SQLite database, a specific table, or a specific column
+    within a specific table. See the following for more details on
+    affinity vs storage class typing within SQLite:
+    https://www.sqlite.org/datatype3.html
+
+    Parameters
+    ----------
+    sql_engine: str | sqlalchemy.engine.base.Engine
+        filename of the SQLite database or existing sqlalchemy engine
+    table_name: str
+        optional specific table name to check within database
+    column_name: str
+        optional specific column name to check within database
+
+    Returns
+    -------
+    bool
+        Returns True if conflicting storage class values were detected
+        in database provided, else returns False.
+    """
+
+    logger.info(
+        (
+            "Determining if SQLite database contains conflicting column "
+            "affinity vs storage class values."
+        )
+    )
+
+    # create an engine
+    engine = engine_from_str(sql_engine)
+
+    # gather columns to be used below
+    columns = collect_columns(engine, table_name, column_name)
+
+    with engine.connect() as connection:
+        for col in columns:
+            # the sql below seeks to efficiently detect existence of values which
+            # do not match the column affinity type (for ex. a string in an integer column).
+            sql_stmt = """
+            SELECT
+                EXISTS(
+                    SELECT 1 FROM {table_name} 
+                    WHERE TYPEOF({col_name}) NOT IN ({col_types})
+                );
+            """
+            # join formatted string for use with sql query in {col_types} var
+            col_types = ",".join(
+                ["'{}'".format(x.lower()) for x in SQLITE_AFF_REF[col[2]]]
+            )
+
+            # there are challenges with using sqlalchemy vars in the same manner as above
+            # so we use format here along with nosec
+            result = connection.execute(
+                sql_stmt.format(
+                    table_name=col[0],
+                    col_name=col[1],
+                    col_types=col_types,
+                )
+            ).fetchall()[0][
```

Comment:
Can you also briefly describe why this is zero indexed twice?

Reply:
SQLAlchemy returns a list of Row objects as tuples when using `fetchall()`. Here and as-is we'd expect `[(int)]` so result[0][0] == int (0 or 1).

Because with these rows we can't receive multiple results we can simplify to use `fetchone()` (removing one index). I'll plan to use this and a named reference, taking advantage of the Row object, to simplify.